lib/src/main/java/org/sonarsource/scanner/lib/internal/JavaRunnerFactory.java

```diff
@@ -0,0 +1,222 @@
+    private final String javaPath;
+
+    JreMetadata(String filename, String sha256, @Nullable String downloadUrl, String id, String javaPath) {
+      super(filename, sha256, downloadUrl);
+      this.id = id;
+      this.javaPath = javaPath;
+    }
+  }
+
+  private static File extractArchive(File cachedFile) {
+    String filename = cachedFile.getName();
+    File destDir = new File(cachedFile.getParentFile(), filename + "_unzip");
+    File lockFile = new File(cachedFile.getParentFile(), filename + "_unzip.lock");
+    if (!destDir.exists()) {
+      try (FileOutputStream out = new FileOutputStream(lockFile)) {
+        FileLock lock = createLockWithRetries(out.getChannel());
+        try {
+          // Recheck in case of concurrent processes
+          if (!destDir.exists()) {
+            File tempDir = Files.createTempDirectory(cachedFile.getParentFile().toPath(), "jre").toFile();
+            extract(cachedFile, tempDir);
+            Files.move(tempDir.toPath(), destDir.toPath());
+          }
+        } finally {
+          lock.release();
+        }
+      } catch (IOException e) {
+        throw new IllegalStateException("Failed to extract archive", e);
+      } finally {
+        deleteQuietly(lockFile.toPath());
+      }
+    }
+    return destDir;
+  }
+
+  private static FileLock createLockWithRetries(FileChannel channel) throws IOException {
+    int tryCount = 0;
+    while (tryCount < 10) {
+      tryCount++;
+      try {
+        return channel.lock();
+      } catch (OverlappingFileLockException ofle) {
+        // ignore overlapping file exception
+      }
+      try {
+        Thread.sleep(200L * tryCount);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+    throw new IOException("Unable to get lock after " + tryCount + " tries");
+  }
+
+  private static void extract(File compressedFile, File targetDir) throws IOException {
+    String extension = compressedFile.getName().substring(compressedFile.getName().lastIndexOf('.') + 1);
+    switch (extension) {
+      case EXTENSION_ZIP:
+        CompressionUtils.unzip(compressedFile, targetDir);
+        break;
+      case EXTENSION_GZ:
+        CompressionUtils.extractTarGz(compressedFile, targetDir);
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported compressed archive extension: " + extension);
+    }
+  }
+
+  static class JreDownloader implements FileCache.Downloader {
+    private final ServerConnection connection;
+    private final JreMetadata jreMetadata;
+
+    JreDownloader(ServerConnection connection, JreMetadata jreMetadata) {
+      this.connection = connection;
+      this.jreMetadata = jreMetadata;
+    }
+
+    @Override
+    public void download(String filename, File toFile) throws IOException {
+      if (StringUtils.isNotBlank(jreMetadata.getDownloadUrl())) {
+        connection.downloadFile(jreMetadata.getDownloadUrl(), toFile.toPath(), false);
```

Comment:
```suggestion
        connection.downloadFromExternalUrl(jreMetadata.getDownloadUrl(), toFile.toPath(), false);
```
be sure to not pass authentication header in this case!

Reply:
Yes there is a boolean param for that